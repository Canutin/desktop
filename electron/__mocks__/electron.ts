const app = {
  dock: {
    hide: jest.fn(),
  },
  getVersion: jest.fn(() => require("../../package.json").version),
  getName: jest.fn(() => "test"),
  getPath: jest.fn(() => "."),
  requestSingleInstanceLock: jest.fn(),
  on: jest.fn(),
  removeAllListeners: jest.fn(),
  quit: jest.fn(),
  get isPackaged() {
    return false;
  },
};

const shell = {
  openExternal: jest.fn(),
};

const dialog = {
  showMessageBoxSync: jest.fn(),
  showSaveDialogSync: jest.fn(() => "/fake/path/to/Canutin.vault"),
  showOpenDialogSync: jest.fn(() => ({
    filePaths: ["/fake/path/to/Canutin.vault"],
  })),
};

const nativeTheme = {
  on: jest.fn(),
  get shouldUseDarkColors() {
    return false;
  },
};

const Tray = jest.fn(() => ({
  on: jest.fn(),
  setContextMenu: jest.fn(),
  setToolTip: jest.fn(),
  setImage: jest.fn(),
  popUpContextMenu: jest.fn(),
}));

const Menu = {
  buildFromTemplate: jest.fn(),
};

module.exports = {
  app,
  shell,
  dialog,
  nativeTheme,
  Menu,
  Tray,
};
