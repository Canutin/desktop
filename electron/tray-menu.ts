import path from "path";
import {
  app,
  Menu,
  MenuItemConstructorOptions,
  shell,
  Tray,
  nativeTheme,
} from "electron";

import Vault from "./vault";
import Server from "./server";

class TrayMenu {
  static readonly OPEN_BROWSER_DELAY = 500;

  static readonly MENU_SERVER_STATUS = "menu-server-status";
  static readonly MENU_SERVER_TOGGLE = "menu-server-toggle";
  static readonly MENU_OPEN_IN_BROWSER = "menu-open-in-browser";
  static readonly MENU_VAULT_PATH = "menu-vault-path";
  static readonly MENU_VAULT_OPEN = "menu-vault-open";

  static readonly ICON_TRAY_IDLE = "canutin-tray-idle";
  static readonly ICON_TRAY_ACTIVE = "canutin-tray-active";
  static readonly ICON_STATUS_POSITIVE = "status-positive";
  static readonly ICON_STATUS_NEGATIVE = "status-negative";

  private isServerRunning: boolean;
  private isAppPackaged: boolean;
  private isMacOs: boolean;

  private tray: Tray;
  private trayIcon: string;
  private vault: Vault;
  server: Server | undefined;

  private menuServerStatus: MenuItemConstructorOptions;
  private menuServerToggle: MenuItemConstructorOptions;
  private menuOpenInBrowser: MenuItemConstructorOptions;
  private menuVaultPath: MenuItemConstructorOptions;
  private menuSwitchVault: MenuItemConstructorOptions;
  private menuPresistentOptions: MenuItemConstructorOptions[];
  private menuSeparator: MenuItemConstructorOptions;
  private menuCurrentTemplate: MenuItemConstructorOptions[];

  constructor(vault: Vault) {
    this.vault = vault;
    this.isServerRunning = false;
    this.isAppPackaged = app.isPackaged || false;
    this.trayIcon = TrayMenu.ICON_TRAY_IDLE;

    // There's a snapshot test that checks the menu template and we don't want
    // to use macOS keyboard shortcuts when generating the template.
    this.isMacOs =
      process.platform === "darwin" && process.env.NODE_ENV !== "test";

    // Menu items (not in order)
    this.menuSeparator = { type: "separator" };
    this.menuServerStatus = {
      label: "Canutin is not running",
      icon: this.getImagePath(TrayMenu.ICON_STATUS_NEGATIVE),
      id: TrayMenu.MENU_SERVER_STATUS,
      enabled: false,
    };
    this.menuServerToggle = {
      label: "Start Canutin",
      click: () => this.toggleServer(),
      id: TrayMenu.MENU_SERVER_TOGGLE,
      visible: false,
    };
    this.menuOpenInBrowser = {
      label: "Open in browser",
      id: TrayMenu.MENU_OPEN_IN_BROWSER,
      accelerator: this.isMacOs ? "Command+T" : "Ctrl+T",
      visible: false,
      click: () => this.openBrowser(),
    };
    this.menuOpenInBrowser = {
      label: "Open in browser",
      id: TrayMenu.MENU_OPEN_IN_BROWSER,
      accelerator: this.isMacOs ? "Command+T" : "Ctrl+T",
      visible: false,
      click: () => this.openBrowser(),
    };
    this.menuVaultPath = {
      label: this.vault?.path ? this.vault.path : "No vault chosen",
      id: TrayMenu.MENU_VAULT_PATH,
      enabled: false,
    };
    this.menuSwitchVault = {
      label: "Switch vault...",
      id: TrayMenu.MENU_VAULT_OPEN,
      accelerator: this.isMacOs ? "Command+S" : "Ctrl+S",
      click: () => this.switchVault(),
    };
    this.menuPresistentOptions = [
      {
        label: "About",
        click: () => app.showAboutPanel(),
      },
      {
        label: "Quit",
        accelerator: this.isMacOs ? "Command+Q" : "Alt+F4",
        click: () => app.quit(),
      },
    ];
    // Menu items (in order)
    this.menuCurrentTemplate = [
      this.menuServerStatus,
      this.menuSeparator,
      this.menuServerToggle,
      this.menuOpenInBrowser,
      this.menuSeparator,
      this.menuVaultPath,
      this.menuSwitchVault,
      this.menuSeparator,
      ...this.menuPresistentOptions,
    ];

    // Set the tray
    this.tray = new Tray(this.getImagePath(TrayMenu.ICON_TRAY_IDLE));
    this.tray.on("click", () => this.tray?.popUpContextMenu()); // Left-click opens the context menu in Windows
    this.tray.setToolTip("Canutin");
    this.tray.setContextMenu(Menu.buildFromTemplate(this.menuCurrentTemplate));

    // Update the tray menu if the OS color theme changes
    nativeTheme.on("updated", () => {
      this.setTrayIcon(this.trayIcon);
    });

    // Start the server
    if (this.vault?.path) {
      this.toggleServer();
      this.isAppPackaged && this.openBrowser(TrayMenu.OPEN_BROWSER_DELAY);
    }
  }

  private switchVault = () => {
    const { vault } = this;
    const isVaultSet = vault.dialog();

    if (isVaultSet && vault.path) {
      this.menuVaultPath.label = vault.path;
      this.updateTray();

      // Starting (or restarting) the server
      if (this.isServerRunning) {
        this.toggleServer(); // Stops the server
        this.toggleServer(); // Starts the server
      } else {
        this.toggleServer(); // Starts the server
        this.openBrowser(TrayMenu.OPEN_BROWSER_DELAY);
      }
    }
  };

  private toggleServer = () => {
    const vaultPath = this.vault?.path;

    if (!this.server && vaultPath) this.server = new Server(vaultPath);

    if (this.server && this.isServerRunning) {
      // Stop the server
      this.server.stop();
      this.menuServerToggle.label = "Start Canutin";
      this.menuServerStatus.label = "Canutin is not running";
      this.menuServerStatus.icon = this.getImagePath(
        TrayMenu.ICON_STATUS_NEGATIVE
      );
      this.menuOpenInBrowser.visible = false;
      this.setTrayIcon(TrayMenu.ICON_TRAY_IDLE);
      this.updateTray();
    } else if (this.server) {
      // Start the server
      this.server.start(vaultPath);
      this.menuServerToggle.visible = true;
      this.menuServerToggle.label = "Stop Canutin";
      this.menuServerStatus.label = "Canutin is running";
      this.menuServerStatus.icon = this.getImagePath(
        TrayMenu.ICON_STATUS_POSITIVE
      );
      this.menuOpenInBrowser.visible = true;
      this.setTrayIcon(TrayMenu.ICON_TRAY_ACTIVE);
      this.updateTray();
    }

    // FIXME:
    // It would be better to check if the server is actually running (or not)
    // instead of blindingly reversing the vaule of `isServerRunning`.
    // REF: https://github.com/Canutin/desktop-2/issues/8
    this.isServerRunning = !this.isServerRunning;
  };

  private updateTray = () => {
    this.tray?.setContextMenu(Menu.buildFromTemplate(this.menuCurrentTemplate));
  };

  private openBrowser(delay: number = 0) {
    const server = this.server;
    // FIXME:
    // To prevent opening the browser before the server is ready we wait.
    // Server boot up time is likely to vary so ideally we would "ping"
    // the server until it's ready and only then open the user's browser.
    server?.url && setTimeout(() => shell.openExternal(server.url), delay);
  }

  private getImagePath(fileName: string) {
    const themeAgnosticIcons = [
      TrayMenu.ICON_STATUS_POSITIVE,
      TrayMenu.ICON_STATUS_NEGATIVE,
    ];

    let theme: string;
    if (themeAgnosticIcons.includes(fileName)) {
      theme = "";
    } else {
      theme = nativeTheme.shouldUseDarkColors ? "-dark" : "-light";
    }

    return this.isAppPackaged
      ? path.join(process.resourcesPath, `assets/${fileName}${theme}.png`)
      : `./resources/${
          process.env.NODE_ENV !== "test" && theme
            ? "assets/dev/dev-"
            : "assets/"
        }${fileName}${theme}.png`;
  }

  private setTrayIcon = (icon: string) => {
    this.trayIcon = icon;
    this.tray.setImage(this.getImagePath(icon));
  };
}

export default TrayMenu;
