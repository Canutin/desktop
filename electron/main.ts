import { app } from "electron";

import TrayMenu from "./tray-menu";
import Vault from "./vault";

process.platform === "darwin" && app.dock.hide(); // Hide dock icon on macOS

let trayMenu: TrayMenu | undefined;
app.whenReady().then(() => {
  const vault = new Vault();

  // Prompt the user to choose a vault if one isn't set yet
  if (!vault.path) vault.openPrompt();

  trayMenu = new TrayMenu(vault);
});

app.on("before-quit", () => {
  trayMenu?.server?.stop();
});
