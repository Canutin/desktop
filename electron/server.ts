import path from "path";
import { app, MessageChannelMain, utilityProcess, UtilityProcess } from "electron";

import { ELECTRON_MESSAGE_SERVER_READY } from "../sveltekit/src/lib/helpers/electron";

class Server {
  static readonly PORT_DEVELOPMENT = "3000";
  static readonly PORT_PRODUCTION = "42069";

  private isAppPackaged: boolean;
  private vaultPath: string;
  private port: string;
  private serverProcess: UtilityProcess | null = null;
  readonly url: string;
  isRunning: boolean;

  constructor(vaultPath: string) {
    this.vaultPath = vaultPath;
    this.isAppPackaged = app.isPackaged;
    this.port = this.isAppPackaged
      ? Server.PORT_PRODUCTION
      : Server.PORT_DEVELOPMENT;
    this.isRunning = false;
    this.url = `http://localhost:${this.port}`;
  }

  start(newVaultPath?: string): Promise<void> {
    return new Promise((resolve, reject) => {
      const HOST = "127.0.0.1";
      const isAppPackaged = this.isAppPackaged;
      const isDev = process.env.NODE_ENV == "development";

      const svelteKitPath = isAppPackaged
        ? path.join(process.resourcesPath, "sveltekit")
        : path.join(__dirname, "../../sveltekit");

      const svelteKitModulePath = isAppPackaged
        ? path.join(svelteKitPath, "index.js")
        : path.join(svelteKitPath, "build", "index.js");

      const { port1 } = new MessageChannelMain();
      this.serverProcess = utilityProcess.fork(svelteKitModulePath, [], {
        env: {
          ...process.env,
          HOST,
          PORT: this.port,
          SVELTEKIT_PATH: svelteKitPath,
          DATABASE_URL: `file:${newVaultPath ? newVaultPath : this.vaultPath}`,
          SHOULD_CHECK_VAULT: "true",
          IS_ELECTRON: "true",
          APP_VERSION: isDev
            ? require("../package.json").version
            : app.getVersion(),
        },
      });

      this.serverProcess.once('spawn', () => {
        this.serverProcess!.postMessage({ message: 'port' }, [port1]);
      });

      this.serverProcess.on('message', async (message) => {
        if (message === ELECTRON_MESSAGE_SERVER_READY) {

          // Send an initial fetch request to initialize the server
          try {
            await fetch(this.url);
          } catch (error) {
            console.error('-> Error initializing server:', error);
            reject(error);
            return;
          }

          this.isRunning = true;
          resolve();
        }
      });

      this.serverProcess.on('exit', (code) => {
        if (code !== 0) reject(new Error(`-> Server process exited with code ${code}`));
      });
    });
  }

  stop() {
    if (this.serverProcess) {
      this.serverProcess.kill();
      this.serverProcess = null;
    }
    this.isRunning = false;
  }
}

export default Server;
