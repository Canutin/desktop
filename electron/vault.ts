import { app, dialog } from "electron";
import path from "path";
import Fs from "fs";
import { exec } from "child_process";
import Store from "electron-store";

class Vault {
  private userSettings: Store;
  path: string | undefined;

  constructor() {
    this.userSettings = new Store();
    this.path = this.getVaultPath();
  }

  saveToUserSettings(vaultPath: string | undefined) {
    this.userSettings.set({ vaultPath: vaultPath || null });
    this.path = vaultPath || undefined;
  }

  dialog = () => {
    const dialogOption = dialog.showMessageBoxSync({
      type: "info",
      title: "Canutin",
      message: "Canutin Vault",
      buttons: ["Create new vault", "Open existing vault", "Do it later"],
      detail:
        "A vault is a file that stores all the data required to run the app",
      cancelId: 2,
    });

    let isVaultSet = false;
    switch (dialogOption) {
      case 0:
        isVaultSet = this.create();
        break;
      case 1:
        isVaultSet = this.load();
        break;
      case 2:
        return false;
    }

    if (isVaultSet) {
      return true;
    } else {
      this.dialog();
    }
  };

  create = () => {
    const newVaultPath = dialog.showSaveDialogSync({
      title: "Canutin",
      defaultPath: "~/Canutin.vault",
      filters: [{ name: "Canutin Vault", extensions: ["vault"] }],
    });

    if (!newVaultPath) return false;

    this.saveToUserSettings(newVaultPath);
    return true;
  };

  load = () => {
    const filePaths = dialog.showOpenDialogSync({
      properties: ["openFile"],
      filters: [{ name: "Canutin Vault", extensions: ["vault"] }],
    });

    if (!filePaths) return false;

    const vaultPath = filePaths[0];
    this.saveToUserSettings(vaultPath);
    return true;
  };

  private getVaultPath = () => {
    const vaultPath = this.userSettings.get("vaultPath") as string | undefined;
    return vaultPath && Fs.existsSync(vaultPath) ? vaultPath : undefined;
  };
}

export default Vault;
