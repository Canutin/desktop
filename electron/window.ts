import { BrowserWindow, BrowserWindowConstructorOptions, app, nativeTheme, screen } from "electron";
import path from "path";

const MIN_WINDOW_WIDTH = 1200;
const MIN_WINDOW_HEIGHT = 768;
const MAX_WINDOW_WIDTH = 1440;
const MAX_WINDOW_HEIGHT = 1200;

const calculateWindowWidth = (displayWidth: number) => {
  const relativeWindowWidth = Math.floor(displayWidth * 0.7);

  if (relativeWindowWidth < MIN_WINDOW_WIDTH) {
    return MIN_WINDOW_WIDTH;
  } else if (relativeWindowWidth > MAX_WINDOW_WIDTH) {
    return MAX_WINDOW_WIDTH;
  } else {
    return relativeWindowWidth;
  }
};

const calculateWindowHeight = (displayHeight: number) => {
  const relativeWindowHeight = Math.floor(displayHeight * 0.7);

  if (relativeWindowHeight < MIN_WINDOW_HEIGHT) {
    return MIN_WINDOW_HEIGHT;
  } else if (relativeWindowHeight > MAX_WINDOW_HEIGHT) {
    return MAX_WINDOW_HEIGHT;
  } else {
    return relativeWindowHeight;
  }
};

export const browserWindowConfig = (): BrowserWindowConstructorOptions => {
  const { width, height } = screen.getPrimaryDisplay().workAreaSize;

  return {
    minWidth: MIN_WINDOW_WIDTH,
    minHeight: MIN_WINDOW_HEIGHT,
    width: calculateWindowWidth(width),
    height: calculateWindowHeight(height),
    backgroundColor: nativeTheme.shouldUseDarkColors ? "#1c1c1c" : "#e3e3e3",
  }
}

export const setLoadingView = (window: BrowserWindow) => {
  const LOADING_HTML = "loading.html";

  if (app.isPackaged) {
    window.loadFile(path.join(process.resourcesPath, 'assets', LOADING_HTML));
  } else {
    window.loadFile(`../../resources/assets/${LOADING_HTML}`);
  }
}
