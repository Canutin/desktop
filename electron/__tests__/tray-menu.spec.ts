import path from "path";
import { BrowserWindow } from "electron";
import Electron, { Menu } from "electron";

import TrayMenu from "../tray-menu";
import Vault from "../vault";
import Server from "../server";

describe("TrayMenu", () => {
  const IMAGE_ASSET = "dev-canutin-tray-active-light";
  // resources/assets/dev/dev-canutin-tray-active-light.png
  const pathToImageAsset = `./resources/assets/dev/${IMAGE_ASSET}.png`;
  const pathToVault = "/fake/path/to/Canutin.vault";
  const resourcesPath = process.resourcesPath; // this is `undefined` in tests
  const mockWindow = new BrowserWindow();
  const vault = new Vault();
  vault.path = pathToVault;

  // FIXME:
  // Tests are passing but there is an error related to `spyPathJoin`:
  //
  // resources/assets/dev/dev-canutin-tray-active-light.png:1
  // ï¿½PNG
  // SyntaxError: Invalid or unexpected token
  //
  // The path to the file is correct. But I think that's failing because we
  // are mocking an Electron module which expects a PNG but the mocked
  // version is maybe expecting something else (?)
  const spyPathJoin = jest
    .spyOn(path, "join")
    .mockReturnValue(pathToImageAsset);
  const spyIsPackaged = jest.spyOn(Electron.app, "isPackaged", "get");

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("tray is set with the correct parameters", () => {
    // FIXME:
    // can't seem to spy on the `toggleServer()` method below:
    // const spyToggleServer = jest.spyOn(
    //   TrayMenu.prototype as any,
    //   "toggleServer"
    // );
    const spyGetImagePath = jest.spyOn(
      TrayMenu.prototype as any,
      "getImagePath"
    );
    const trayMenu = new TrayMenu(vault, mockWindow);
    const tray = trayMenu["tray"];
    expect(trayMenu["menuCurrentTemplate"]).toMatchSnapshot();
    expect(spyGetImagePath).toHaveBeenCalledWith(TrayMenu.ICON_TRAY_IDLE);
    expect(tray.on).toHaveBeenCalledWith("click", expect.any(Function));
    expect(tray.setToolTip).toHaveBeenCalledWith("Canutin");
    expect(tray.setContextMenu).toHaveBeenCalledWith(
      Menu.buildFromTemplate(trayMenu["menuCurrentTemplate"])
    );
    expect(trayMenu["vault"]["path"]).toBe(pathToVault);
    // expect(spyToggleServer).toHaveBeenCalled();
  });

  describe("develoment environment", () => {
    spyIsPackaged.mockReturnValue(false);
    const trayMenu = new TrayMenu(vault, mockWindow);

    test("browser opens to the server url", () => {
      expect(trayMenu["server"]?.url).toBe(
        `http://localhost:${Server.PORT_DEVELOPMENT}`
      );
    });

    test("path to image assets", () => {
      spyPathJoin.mockClear();
      const imagePath = trayMenu["getImagePath"](IMAGE_ASSET);
      expect(imagePath).toBe(`./resources/assets/${IMAGE_ASSET}-light.png`);
      expect(spyPathJoin).not.toBeCalled();
    });
  });

  describe("production environment", () => {
    spyIsPackaged.mockReturnValue(true);
    const trayMenu = new TrayMenu(vault, mockWindow);

    test("browser opens to the server url", () => {
      expect(trayMenu["server"]?.url).toBe(
        `http://localhost:${Server.PORT_PRODUCTION}`
      );
    });

    test("path to image assets", () => {
      const imagePath = trayMenu["getImagePath"](IMAGE_ASSET);
      expect(imagePath).toBe(pathToImageAsset);
      expect(spyPathJoin).toHaveBeenLastCalledWith(
        resourcesPath,
        `assets/${IMAGE_ASSET}-light.png`
      );
    });
  });

  test("tray icon is loaded with the correct theme color", () => {
    const iconTrayIdle = TrayMenu.ICON_TRAY_IDLE;
    const iconTrayActive = TrayMenu.ICON_TRAY_ACTIVE;
    const iconStatusPositive = TrayMenu.ICON_STATUS_POSITIVE;
    const iconStatusNegative = TrayMenu.ICON_STATUS_NEGATIVE;

    const spyShouldUseDarkColors = jest.spyOn(
      Electron.nativeTheme,
      "shouldUseDarkColors",
      "get"
    );
    const trayMenu = new TrayMenu(vault, mockWindow);

    expect(trayMenu["trayIcon"]).toBe(iconTrayActive);
    expect(spyPathJoin).toHaveBeenCalledWith(
      resourcesPath,
      `assets/${iconTrayIdle}-light.png`
    );
    expect(spyPathJoin).toHaveBeenLastCalledWith(
      resourcesPath,
      `assets/${iconTrayActive}-light.png`
    );

    spyShouldUseDarkColors.mockReturnValue(true);
    trayMenu["setTrayIcon"](iconTrayActive);
    expect(spyPathJoin).toHaveBeenLastCalledWith(
      resourcesPath,
      `assets/${iconTrayActive}-dark.png`
    );

    // Positive and negative icons are theme agnostic
    expect(spyPathJoin).toHaveBeenCalledWith(
      resourcesPath,
      `assets/${iconStatusPositive}.png`
    );
    expect(spyPathJoin).not.toHaveBeenCalledWith(
      resourcesPath,
      `assets/${iconStatusPositive}-light.png`
    );
    expect(spyPathJoin).not.toHaveBeenCalledWith(
      resourcesPath,
      `assets/${iconStatusPositive}-dark.png`
    );
    expect(spyPathJoin).toHaveBeenCalledWith(
      resourcesPath,
      `assets/${iconStatusNegative}.png`
    );
    expect(spyPathJoin).not.toHaveBeenCalledWith(
      resourcesPath,
      `assets/${iconStatusNegative}-light.png`
    );
    expect(spyPathJoin).not.toHaveBeenCalledWith(
      resourcesPath,
      `assets/${iconStatusNegative}-dark.png`
    );

    // TODO: add test for when the OS changes theme color `nativeTheme.on("updated"...`
  });

  // Disabled tests
  // REF https://github.com/Canutin/desktop-2/issues/6
  test.skip("server can be toggled on/off", () => {
    // TODO
  });

  test.skip("vault can be opened", () => {
    // FIXME:
    // half of the assertions in this test don't work.
    const spyElectronDialogSync = jest.spyOn(
      Electron.dialog,
      "showOpenDialogSync"
    );
    const spyUpdateTray = jest.spyOn(TrayMenu.prototype as any, "updateTray");
    const trayMenu = new TrayMenu(vault, mockWindow);
    trayMenu["switchVault"];
    expect(trayMenu["menuVaultPath"].label).not.toBe(pathToVault);
    expect(spyElectronDialogSync).toHaveBeenCalledWith({
      properties: ["openFile"],
      filters: [{ name: "Canutin Vault", extensions: ["vault"] }],
    });
    expect(spyElectronDialogSync).toHaveReturnedWith({
      canceled: false,
      filePaths: [pathToVault],
    });
    expect(trayMenu["menuVaultPath"].label).toBe(pathToVault);
    expect(trayMenu["vault"].saveToUserSettings).toHaveBeenCalledWith(
      pathToVault
    );
    expect(spyUpdateTray).toHaveBeenCalled();
  });
});
