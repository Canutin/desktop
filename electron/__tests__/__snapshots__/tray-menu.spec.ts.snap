// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrayMenu tray is set with the correct parameters 1`] = `
[
  {
    "enabled": false,
    "icon": "/path/to/fake/assets/canutin-tray-icon.png",
    "id": "menu-server-status",
    "label": "Canutin is running",
  },
  {
    "type": "separator",
  },
  {
    "click": [Function],
    "id": "menu-server-toggle",
    "label": "Stop Canutin",
    "visible": true,
  },
  {
    "type": "separator",
  },
  {
    "enabled": false,
    "id": "menu-vault-path",
    "label": "/fake/path/to/Canutin.vault",
  },
  {
    "accelerator": "Ctrl+S",
    "click": [Function],
    "id": "menu-vault-open",
    "label": "Switch vault...",
  },
  {
    "type": "separator",
  },
  {
    "click": [Function],
    "label": "About",
  },
  {
    "accelerator": "Alt+F4",
    "click": [Function],
    "label": "Quit",
  },
]
`;
